### PR TITLE
[DO NOT MERGE]Decompose MultiheadAttention module

### PR DESCRIPTION
**Pull request by Enealor**
Decomposes the function `multi_head_attention_forward` into three functions to address #34573 and introduces three new modules that use these functionals. I ultimately decided to do a minimalist approach, which adds new functions and reuses them as much as possible. The breakdown is as follows:
* `multi_head_attention_in_projection` - handles the projection prior to applying attention.
* `scaled_dot_product_attention` - handles turning the projected query, key and value into an updated query. It also outputs attention weights.
* `multi_head_attention_out_projection` - handles the transformation from projected query to embedded query.

Each function has an associated module that store the parameters for ease of use.

Note: The argument for each has the parameter `num_heads` appear before the weight and bias. I chose to do this as I felt weight and bias should be grouped, and that bias should be an optional argument. An alternative approach is to give `num_heads` a default argument so that it is also optional (and then move it to after giving the weights). On the third hand, it's not strictly necessary as it can be worked out from the shape of other tensors in *some* cases. I decided to make it required on each for consistency.

I tested the code to verify that different argument paths would give the same answer. Below is the script I used to do so:
```
from random import randint
import torch
import torch.nn.functional as F

tgt_len = randint(5, 15)
src_len = randint(5, 15)
bsz = randint(20, 30)
head_dim = randint(4, 10)
num_heads = randint(4, 10)
proj_dim = head_dim * num_heads
embed_dim = proj_dim    # Necessary for current form

query = torch.randn((tgt_len, bsz, embed_dim))
key = torch.randn((src_len, bsz, embed_dim))
value = torch.randn((src_len, bsz, embed_dim))

w_q = torch.rand((proj_dim, embed_dim))
b_q = torch.rand((proj_dim,))
q = F.multi_head_attention_in_projection(query, num_heads, w_q, b_q)

w_k = torch.rand((proj_dim, embed_dim))
b_k = torch.rand((proj_dim,))
k = F.multi_head_attention_in_projection(key, num_heads, w_k, b_k)

w_v = torch.rand((proj_dim, embed_dim))
b_v = torch.rand((proj_dim,))
v = F.multi_head_attention_in_projection(value, num_heads, w_v, b_v)

attn, attn_weight = F.scaled_dot_product_attention(q, k, v, num_heads, False, 0)

w_out = torch.rand((embed_dim, proj_dim))
b_out = torch.rand((embed_dim,))

new_query_1 = F.multi_head_attention_out_projection(attn, num_heads, w_out, b_out)
attn_weight_1 = attn_weight.reshape(bsz, num_heads, tgt_len, src_len).sum(dim=1) / num_heads
new_query_2, attn_weight_2 = F.multi_head_attention_forward(
    query, key, value, embed_dim, num_heads, None, torch.cat([b_q, b_k, b_v]),
    None, None, False, 0, w_out, b_out, use_separate_proj_weight=True,
    q_proj_weight=w_q, k_proj_weight=w_k, v_proj_weight=w_v)
new_query_3, attn_weight_3 = F.multi_head_attention_forward(
    query, key, value, embed_dim, num_heads, torch.cat([w_q, w_k, w_v]),
    torch.cat([b_q, b_k, b_v]), None, None, False, 0, w_out, b_out)

assert torch.all(new_query_1 == new_query_2), 'Wrong query output'
assert torch.all(attn_weight_1 == attn_weight_2), 'Wrong attention weights'
assert torch.all(new_query_1 == new_query_3), 'Wrong query output'
assert torch.all(attn_weight_1 == attn_weight_3), 'Wrong attention weights'
```

Verified that the functions support JIT tracing using the following script (variables are from above):
```
mha_in = lambda q, w, b: F.multi_head_attention_in_projection(q, num_heads, w, b)
sdp = lambda q, k, v: F.scaled_dot_product_attention(q, k, v, num_heads, False, 0)
mha_out = lambda a, w, b: F.multi_head_attention_out_projection(a, num_heads, w, b)

torch.jit.trace(mha_in, (query, w_q, b_q))
torch.jit.trace(sdp, (q, k, v))
torch.jit.trace(mha_out, (attn, w_out, b_out))
```
TODO:
- [ ] Update type hints
- [x] Compile and verify documentation
- [ ] Write unit tests
- [x] Test JIT support
- [ ] Test quantization support

### CODE REVIEW

**Comment by Enealor** [comment]
@zhangguanheng66 @cpuhrsch

**Review by zhangguanheng66 — CHANGES_REQUESTED** [review]
For the initial work, I think it's better to not touch the `nn.functional.multi_head_attention_forward` function because it will affect too many downstream users.

File: torch/nn/functional.py
Comment (zhangguanheng66) [inline torch/nn/functional.py]: This function is same for query/key/value. We can add a single in_projection function which work on them individually. See the function [here](https://github.com/pytorch/pytorch/pull/34719/files#diff-c5db6ae353758a64f3e73dd758728cb7R378)

**Review by zhangguanheng66 — CHANGES_REQUESTED** [review]

**Review by zhangguanheng66 — COMMENTED** [review]
If you like, you can check out my PR https://github.com/pytorch/pytorch/pull/34719, add `attn_mask`/`key_padding` arguments, test JIT support, update docs, add unit test, and move the code to `functional` which will be called by the forward function.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: And?

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: That is what I have done here.

**Comment by Enealor** [comment]
I don't see the benefit of that approach, as it seems likely to miss the challenges of decomposing the functional and require retooling. The impact on downstream users when changing the functional should be zero, as I'm not changing the functionality; I am only reusing code where I have moved it elsewhere.

This PR still needs work as some of the tests are failing, but it is in a form that is ready for actual feedback. If your interest is just in getting the new modules, then I can close this until a better time.

**Comment by zhangguanheng66** [comment]
> I don't see the benefit of that approach, as it seems likely to miss the challenges of decomposing the functional and require retooling. The impact on downstream users when changing the functional should be zero, as I'm not changing the functionality; I am only reusing code where I have moved it elsewhere.
> 
> This PR still needs work as some of the tests are failing, but it is in a form that is ready for actual feedback. If your interest is just in getting the new modules, then I can close this until a better time.

The idea is to provide users some building blocks so users can build MHA with more flexibility. I agree that eventually we will merge those code to the `multi_head_attention_forward` function. Nevertheless, we need those new modules as well as the decomposed functionals first.

**Comment by Enealor** [comment]
Okay. I am currently working on incorporating the modules based on your PR. I've made changes to the module documentation to make it align with the description I used. 

I can revert the changes to changes to the original `multi_head_attention_forward` for now, and test the new modules and functionals against the current. I can start adding tests if there aren't changes we should address first.

**Comment by zhangguanheng66** [comment]
> Okay. I am currently working on incorporating the modules based on your PR. I've made changes to the module documentation to make it align with the description I used.
> 
> I can revert the changes to changes to the original `multi_head_attention_forward` for now, and test the new modules and functionals against the current. I can start adding tests if there aren't changes we should address first.

Sounds good to me. Let's keep this PR away from the original `multi_head_attention_forward`. Then, I will take a close look at the PR and conduct some benchmark work. There is a questions that where we should go with a single large project matrix for query/key/value or three separate matrix. We need to prove that there is no performance regression for those new building blocks.

**Comment by Enealor** [comment]
The test code I have above verifies that chaining `multi_head_attention_in_projection`, `scaled_dot_product_attention` and `multi_head_attention_out_projection` produces the same results for two of the code paths. However, the tests failed due to incorrect results.

After investigating the failed tests, I've found that it produces incorrect results for the self-attention path and the encoder-decoder path. The difference might be a result of numerical accuracy, but I do not entirely understand the cause. 

In particular, the following attempt at replicating the self-attention path produces results where the norm of the difference is around 0.0001 and goes down when using `double`.
```
w_a = torch.cat([w_q, w_k, w_v])
key = value = query
q1 = F.multi_head_attention_in_projection(query, num_heads, w_q, b_q)
k1 = F.multi_head_attention_in_projection(key, num_heads, w_k, b_k)
v1 = F.multi_head_attention_in_projection(value, num_heads, w_v, b_v)

q2, k2, v2 = F.linear(query, w_a, b_a).chunk(3, -1)
q2 = q2.reshape(tgt_len, bsz * num_heads, head_dim).transpose(0, 1)
k2 = k2.reshape(-1, bsz * num_heads, head_dim).transpose(0, 1)
v2 = v2.reshape(-1, bsz * num_heads, head_dim).transpose(0, 1)
print("{0:.5}".format(torch.norm(q2 - q1)))
print("{0:.5}".format(torch.norm(k2 - k1)))
print("{0:.5}".format(torch.norm(v2 - v1)))
```
(Values are chosen randomly, but differences are consistently low magnitude.)

I had suspected it was related to `chunk`, but I have ruled that out. Mathematically, chunking after works out to the same as using linear with each weight separately. As that is the primary difference, I am now suspecting numerical inaccuracy.

**Comment by zhangguanheng66** [comment]
Might be relevant but not sure? https://github.com/pytorch/pytorch/issues/33841

**Comment by Enealor** [comment]
Thanks for link. I think it is the same root issue; that issue has a follow-up (#34060) that mirrors what I'm seeing here. With that in mind, I think the current form is functionally complete. There will be a discrepancy in output due to the linked issue, but that is outside the scope of this PR.

File: torch/nn/modules/transformer.py
Comment (Enealor) [inline torch/nn/modules/transformer.py]: Should we include the ability to enable and disable a bias parameter? Would be simple enough to add an optional `bias` parameter.

File: torch/nn/modules/transformer.py
Comment (Enealor) [inline torch/nn/modules/transformer.py]: I basically reused the code from `nn.Linear` for weights and initialization. This offers the flexibility to choose a different initialization. Suggestions?

File: torch/nn/modules/transformer.py
Comment (Enealor) [inline torch/nn/modules/transformer.py]: I'm starting to think it might be best to make a single module that does `q`, `k` and `v` all at once. It reduces flexibility, but it will be a more accessible interface. People that want the flexibility could use the `functional` as well to construct their own module. Thoughts on a split approach?

File: torch/nn/modules/transformer.py
Comment (Enealor) [inline torch/nn/modules/transformer.py]: Include bias? I'll use the same answer as above.

File: torch/nn/functional.py
Comment (zhangguanheng66) [reply torch/nn/functional.py]: Yes. `bmm` is limited to the 3-D tensors. With `matmul`, it removes this limitation. `query`, `key`, `value` may be in the shape of (N, H, S, E), where N is the batch, H is the head dimension, S is the sequence length, and E is the feature. With `bmm`, we have to combine `H` and `N` dimensions together as `(N*H, S, E)`.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Actually I don't have a strong opinion against this. @fmassa Do you have a preference?

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: I would not expose this to users for now.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Not for now. It's always easier to add codes than removing.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: Same as above.

File: torch/nn/modules/transformer.py
Comment (Enealor) [reply torch/nn/modules/transformer.py]: Sounds good.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: Changed it to matmul. I can change the docs to indicate either size is usable.

File: torch/nn/modules/transformer.py
Comment (Enealor) [reply torch/nn/modules/transformer.py]: I agree. I meant to ask if there is a better initialization to be selecting. It's not a high priority, but it's something I am curious about.

**Comment by Enealor** [comment]
Added example usage to documentation. Realized while making them that I don't like the nomenclature I created. The original motivation is that I was mimicking the names of the weights with consideration to different embedding sizes. I'm going to change the initialization signature of the projections to `__init__(embed_dim, num_heads, head_dim=None)`, where `head_dim = embed_dim // num_heads` when not set. This will let them have a consistent interface that is easier to process; I kept mixing up dimensions while writing examples.

**Review by zhangguanheng66 — COMMENTED** [review]
Thanks for the work. We need some unit test. One of them should compare the results of building blocks with current MHA functional/module. Another one should make sure the JIT support with building blocks. Here are some references

https://github.com/pytorch/pytorch/blob/fa5bc9fa2e995432986aa374f70309298f3abd50/test/test_nn.py#L4708

https://github.com/pytorch/pytorch/blob/fa5bc9fa2e995432986aa374f70309298f3abd50/test/test_jit.py#L17140

File: torch/nn/functional.py
Comment (zhangguanheng66) [inline torch/nn/functional.py]: I think we should return `attn_output_weights` before `dropout`.

File: torch/nn/functional.py
Comment (zhangguanheng66) [inline torch/nn/functional.py]: I have a PR to add bool tensor support for `attn_mask` and eventually, we will switch to bool tensor for both `attn_mask` and `key_padding_mask`. For those building blocks, I think we should start with bool tensor for the masks.
https://github.com/pytorch/pytorch/pull/33763

Comment (Enealor) [reply torch/nn/functional.py]: I can work on adding that once the PR lands.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: I considered that. I chose to return it after dropout as that will be used to compute the `attn_output`. Furthermore, if a user wants the weights without dropout, they can use `training=False`. If we specifically want the result without dropout though, I can use an auxiliary variable.

File: torch/nn/functional.py
Comment (zhangguanheng66) [reply torch/nn/functional.py]: No worry for that PR. I will make sure it land. 
For this PR, I think it makes sense to use bool tensor for `attn_mask` and `key_padding_mask`.

File: torch/nn/functional.py
Comment (zhangguanheng66) [reply torch/nn/functional.py]: I think it's more intuitive to return the full `attn_output_weights` because users may want to take a look at the output weights to understand the score between query and key. You can do something like this
```
attn_output = torch.matmul(dropout(attn_output_weights, p=dropout_p, training=training), v)
```

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [inline torch/nn/modules/transformer.py]: Mentioned by @fmassa , we should unify the input shape of those three building blocks by sequence-dimension first.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: Okay. I have also included support for a boolean `attn_mask`. `key_padding_mask` was already a boolean tensor.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: You can have either batch-first or seq-first input for the three building blocks. In this case, IMO, it makes sense to have seq-first to be consistent with MHA. How do you feel about it?

File: torch/nn/functional.py
Comment (zhangguanheng66) [reply torch/nn/functional.py]: I think it makes sense to support only bool tensor for `attn_mask` and `key_mask`. We have both float and bool support for `attn_mask` in MHA because the `attn_mask` was initially implemented as float tensor. We don't need to do that again for those new building blocks as long as we make it clear in the doc.

File: torch/nn/modules/transformer.py
Comment (Enealor) [reply torch/nn/modules/transformer.py]: I understand now. I can have it stay consistent across layers so that the first dimension is the sequence length. As a follow-up, how do you feel about using the shape `(S, N, H, P / H)` instead? I think it would be easier to work with the functional in this form as picking out an example from the batch would not require reshaping or messy slicing.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: Okay. I have an update that changes the documentation and removes the type-dependent behavior.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: I agree with this. I think the reason behind the three-dimension query/key/value tensors is the `bmm` ops used in the MHA. Since we are not constrained by this limitation, we can move forward with the four-dimension tensors.

I plan to pull more people to take a look at the those building blocks (next week?) once the PR is matured.

File: torch/nn/modules/transformer.py
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Do we story around broadcasting for this?

File: docs/source/nn.rst
Comment (cpuhrsch) [inline docs/source/nn.rst]: Is it necessary for this to be specific to MultiheadAttention?

File: torch/nn/modules/transformer.py
Comment (cpuhrsch) [inline torch/nn/modules/transformer.py]: Is this a concept more generic that this? I.e. something that is used in other Attention mechanisms as well etc.

**Comment by cpuhrsch** [comment]
It might make sense to focus this diff on ScaledDotProduct for now since the in and out projections are relatively generic and we might want to name them something else or land something more general.

**Comment by cpuhrsch** [comment]
We could think of this in terms of a wider effort to introduce a collection of attention functions, which are then used to assemble various Attention mechanisms.

Just like George enumerated them already, there is quite a few of them.

For that I'd then make sure they support
- broadcasting
- masking
- quantization
- JIT
and eventually a path to an efficient native C++ implementation.

File: torch/nn/modules/transformer.py
Comment (zhangguanheng66) [reply torch/nn/modules/transformer.py]: The counterargument is that with both in/out projection modules users will have the JIT/quantization support, which may not be straightforward.
For some users who are interested to implement a custom attention layers, they don't need to re-write the in/out projection layers. For example, if they want to change the score func in the dot product layer, they can still re-use those in/out projection modules without any changes.

Comment (cpuhrsch) [reply torch/nn/modules/transformer.py]: I agree that this is useful, but I suspect that the name doesn't do it's wider applicability justice.

File: torch/nn/functional.py
Comment (raghuramank100) [inline torch/nn/functional.py]: Why is this part needed?

File: torch/nn/functional.py
Comment (raghuramank100) [inline torch/nn/functional.py]: Have we verified that calling this module thrice is as efficient as calling it with a single concatenated weight generating q,k and v?

File: torch/nn/functional.py
Comment (raghuramank100) [inline torch/nn/functional.py]: Are all of these functions scriptable?

Comment (zhangguanheng66) [reply torch/nn/functional.py]: We should have the JIT support for all the three new functions.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py]: We have not. That is something that @zhangguanheng66 offered to do. It's not guaranteed that this module is even going to persist in this form at the moment, so it's a premature to do so.

File: torch/nn/functional.py
Comment (Enealor) [reply torch/nn/functional.py:4076]: I have no idea. It's included with the rest, and I assumed it was part of the `jit` package which I'm not fully familiar with.

**Comment by Enealor** [comment]
I added a test for the numerical accuracy for the scaled dot product. I also commented out the MHA projections since it sounds like that isn't wanted. However, I am getting conflicting directions on what *is* wanted.
* Masks support only booleans because that's what I was directed to do. If you have a specific type I should include, please specify which.
* Is there a final word on including the projections or just removing them? While they aren't difficult, I understood their inclusion to be part of the goal. If they are not wanted, I can delete them.

I can work on including the JIT tests or other tests when I have a definite goal. At the moment, it feels like I'm being asked to work on things that I may just then remove.

**Comment by zhangguanheng66** [comment]
> I added a test for the numerical accuracy for the scaled dot product. I also commented out the MHA projections since it sounds like that isn't wanted. However, I am getting conflicting directions on what _is_ wanted.
> 
> * Masks support only booleans because that's what I was directed to do. If you have a specific type I should include, please specify which.
> * Is there a final word on including the projections or just removing them? While they aren't difficult, I understood their inclusion to be part of the goal. If they are not wanted, I can delete them.
> 
> I can work on including the JIT tests or other tests when I have a definite goal. At the moment, it feels like I'm being asked to work on things that I may just then remove.

I pulled the work to torchtext because the library is more research related. If you like, you could continue your work there (JIT test). I added a numeric test there and will work on a Multihead contrainer, which accept different attention functions. Your work has been appreciated and we plan to merge it eventually.
For the attention mask, I think we should just support byte tensor. And we prefer a simple API with minimized arguments so we don't need `add_attn` for the building blocks.

**Pull request closed without merging**
